### PR TITLE
Year in music and a whole pile of other general development

### PR DESCRIPTION
**Pull request by mayhem**
This PR is too large and I apologize for that. It originally started off for the 2020 year in review work, but then languished until now. And in creating the playlist for this years year review, I needed to create many more elements in order to get the job accomplished.

Some of the work in this PR:

- Create top discoveries patch
- Create top recordings patch
- Create top missed recordings
- Add BPM lookup element
- Add Mood lookup element
- Add playlist shaper that orders a playlist by BPM, from low to high to low again.
- Add a playlist shuffle element
- Add a playlist redundancy reducer element
- Add a playlist duplicate reducer that removes duplicates based on artist/track name for use with Annoy
- Add type checking of the pipeline so that input/output mismatching pipelines cannot be created.
- Add command line support to override playlist name/desc
- Add a patch to read a file with MBIDs and make a playlist from them.

This PR does not address tests -- that will come in the next version. But lets aim to merge this ASAP so that it does not continue to get bigger.

### CODE REVIEW

**Review by amCap1712 — APPROVED** [review]
Few comments. otherwise LGTM. More docs and test will also help but we already decided to do that in follow up PR so that is fine.

File: troi/print_recording.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+            self.print_listen_count = True
+        else:
+            self.print_listen_count = False
+
+        if "bpm" in recording.acousticbrainz:
+            self.print_bpm = True
+        else:
+            self.print_bpm = False
+
+        if "moods" in recording.acousticbrainz:
```
[the remark this answers is not in the document]
Comment (mayhem) [reply troi/print_recording.py:35]: Yeah, I'm not really a fan of that form, TBH.

File: troi/cli.py [hunk 2]
```diff
@@ -56,7 +58,15 @@ def playlist(patch, debug, echo, save, token, args, created_for):
     try:
         playlist = troi.playlist.PlaylistElement()
         playlist.set_sources(pipeline)
+        print("Troi playlist generation starting...")
         playlist.generate()
+
+        if name:
+            playlist.playlists[0].name = name
+        if desc:
+            playlist.playlists[0].descripton = description
```
[the remark this answers is not in the document]
Comment (mayhem) [reply troi/cli.py:67]: whops, clearly not used that.

File: troi/acousticbrainz/mood_lookup.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+        max_items_per_call = 25
+        for rec_chunk in chunks(recordings, max_items_per_call):
+            mbids = [ r.mbid for r in rec_chunk ]
+            r = requests.post(self.SERVER_URL + ";".join(mbids))
+            if r.status_code != 200:
+                raise PipelineError("Cannot fetch moods from AcousticBrainz: HTTP code %d" % r.status_code)
+
+            data = r.json()
+            output = []
```
[the remark this answers is not in the document]
Comment (mayhem) [reply troi/acousticbrainz/mood_lookup.py:56]: Oh wow, that might explain the low coverage rate I saw for AB. Will retry that!

File: troi/playlist.py [hunk 4]
```diff
@@ -176,3 +171,137 @@ def submit(self, token, created_for):
...
+        return [Playlist]
+
+    def bpm_sawtooth_sort(self, recordings):
+        try:
+            sorted_recs = sorted(recordings, key=lambda rec: rec.acousticbrainz["bpm"])
+        except AttributeError:
+            raise RuntimeError("acousticbrainz.bpm not set for recording in playlist PBM sort.")
+
+        index = sorted_recs.index(max(sorted_recs, key=lambda rec: rec.acousticbrainz["bpm"]))
+        while index >= 0:
```
[the remark this answers is not in the document]
Comment (mayhem) [reply troi/playlist.py:270]: +1

File: troi/playlist.py [hunk 5]
```diff
@@ -69,14 +72,17 @@ class PlaylistElement(Element):
...
         super().__init__()
         self.playlists = []
+        self.print_recording = PrintRecordingList()
 
     @staticmethod
     def inputs():
         return [Recording, Playlist]
 
+    def __str__(self):
+        print(str(self.playlists))
```
[the remark this answers is not in the document]
Comment (mayhem) [reply troi/playlist.py:82]: d'oh.

File: troi/patches/world_trip.py [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+        if inputs["sort"] == "longitude":
+            latitude = False
+            lat_string = "North -> South"
+        else:
+            latitude = True
+            lat_string = "West -> East"
+
+        trip = WorldTripElement(inputs['continent'], latitude)
+        pl_maker = troi.playlist.PlaylistMakerElement(self.NAME % (inputs["continent"], lat_string),
+                                                      self.DESC % (quote(inputs["continent"]), lat_string), max_items=5000)
```
[the remark this answers is not in the document]
Comment (mayhem) [reply troi/patches/world_trip.py:173]: Thats clearly not the right way to fix that... :)

[the remark this answers is not in the document]
Comment (mayhem) [reply troi/patches/world_trip.py:173]: Oh, I actually did already improve that. Not passing max_tracks does not truncate the list...

File: troi/patches/ab_similar_recordings.py [hunk 7]
```diff
@@ -3,6 +3,7 @@
 import troi
 import troi.patch
 import troi.filters
+import troi.utils
```
[the remark this answers is not in the document]
Comment (mayhem) [reply troi/patches/ab_similar_recordings.py:6]: +1

**Pull request merged**